sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala

```diff
@@ -121,18 +122,36 @@ class MicroBatchExecution(
       // v2 source
       case r: StreamingDataSourceV2Relation => r.stream
     }
-    uniqueSources = sources.distinct.map {
-      case source: SupportsAdmissionControl =>
-        val limit = source.getDefaultReadLimit
-        if (trigger == OneTimeTrigger && limit != ReadLimit.allAvailable()) {
-          logWarning(s"The read limit $limit for $source is ignored when Trigger.Once() is used.")
-          source -> ReadLimit.allAvailable()
-        } else {
-          source -> limit
-        }
-      case other =>
-        other -> ReadLimit.allAvailable()
-    }.toMap
+    uniqueSources = triggerExecutor match {
+      case _: SingleBatchExecutor =>
+        sources.distinct.map {
+          case s: SupportsAdmissionControl =>
+            val limit = s.getDefaultReadLimit
+            if (limit != ReadLimit.allAvailable()) {
+              logWarning(
+                s"The read limit $limit for $s is ignored when Trigger.Once is used.")
+            }
+            s -> ReadLimit.allAvailable()
+          case s =>
+            s -> ReadLimit.allAvailable()
+        }.toMap
+
+      case _: MultiBatchExecutor =>
+        sources.distinct.map {
+          case s: SupportsTriggerAvailableNow => s
+          case s: Source => new AvailableNowSourceWrapper(s)
+          case s: MicroBatchStream => new AvailableNowMicroBatchStreamWrapper(s)
+        }.map { s =>
+          s.prepareForTriggerAvailableNow()
+          s -> s.getDefaultReadLimit
```

Comment:
Should we log warning if it is `getDefaultReadLimit` is `ReadLimit.allAvailable()` for `AvailableNowTrigger`?